internals/controllers/rolebinding/rolebinding_cleanup.go

```diff
@@ -0,0 +1,108 @@
+package controllers
+
+import (
+	"fmt"
+	danav1 "github.com/dana-team/hns/api/v1"
+	"github.com/dana-team/hns/internals/utils"
+	"github.com/go-logr/logr"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// cleanUp takes care of clean-up related operations that need to be done when
+// a roleBinding relevant to HNS is deleted
+func (r *RoleBindingReconciler) cleanUp(rbObject *utils.ObjectContext, sns *utils.ObjectContextList) error {
+	ctx := rbObject.Ctx
+	logger := log.FromContext(ctx)
+	logger.Info("cleaning up roleBinding")
+
+	rbName := rbObject.Object.GetName()
+	rbNamespace := rbObject.Object.GetNamespace()
+
+	if !IsServiceAccount(rbObject.Object) {
+		if err := deleteSubjectsFromHNSViewClusterRoleBinding(rbObject); err != nil {
+			return fmt.Errorf("failed to delete subjects from roleBinding '%s' to HNS View ClusterRoleBinding: "+err.Error(), rbName)
+		}
+		logger.Info("successfully deleted subjects from roleBinding to HNS View ClusterRoleBinding", "roleBinding", rbName)
+	}
+
+	if err := deleteRoleBindingsInSnsList(rbObject, sns); err != nil {
+		return fmt.Errorf("failed to delete RoleBinding in every child of namespace '%s': "+err.Error(), rbNamespace)
+	}
+	logger.Info("successfully deleted RoleBinding in every child of namespace", "roleBinding namespace", rbNamespace)
+
+	if err := deleteRBFinalizer(rbObject); err != nil {
+		return fmt.Errorf("failed to delete finalizer to roleBinding '%s': "+err.Error(), rbName)
+	}
+	logger.Info("successfully deleted finalizer to roleBinding", "roleBinding", rbName)
+
+	return nil
+}
+
+// deleteSubjectsFromHNSViewClusterRoleBinding deletes any subjects existing in the reconciled
+// roleBinding from the relevant HNS View CRB
+func deleteSubjectsFromHNSViewClusterRoleBinding(rbObject *utils.ObjectContext) error {
+	nsHNSViewClusterRoleBinding, err := getNamespaceHNSViewCRB(rbObject)
+	if err != nil {
+		return err
+	}
+
+	hnsViewSubjects := nsHNSViewClusterRoleBinding.Object.(*rbacv1.ClusterRoleBinding).Subjects
+
+	var newSubjects []rbacv1.Subject
+	roleBindingSubjects := rbObject.Object.(*rbacv1.RoleBinding).Subjects
+
+	for _, subject := range hnsViewSubjects {
+		if !isSubjectInSubjects(roleBindingSubjects, subject) {
+			newSubjects = append(newSubjects, subject)
+		}
+	}
+
+	return updateNamespaceHNSViewCRBSubjects(nsHNSViewClusterRoleBinding, newSubjects)
+}
+
+// deleteRoleBindingsInSNSList deletes the RoleBinding in every namespace linked to a subnamespace
+// that is given as input in the snsList slice
+func deleteRoleBindingsInSnsList(rbObject *utils.ObjectContext, snsList *utils.ObjectContextList) error {
```

Comment:
I need explanation about this function but for sure its to long a need to use function for single creation and single deletion

Reply:
There was an error in this function, left over from copy-paste. Only the delete part should stay